openlibrary/core/models.py

```diff
@@ -593,6 +597,108 @@ def get_redirect_chain(work_key: str) -> list:
                 resolved_key = thing.key
         return redirect_chain
 
+    @classmethod
+    def resolve_redirect_chain(cls, work_key: str, test: bool = False) -> dict[str, Any]:
+        summary: dict[str, Any] = {
+            'key': work_key,
+            'redirect_chain': [],
+            'resolved_key': None
+        }
+        redirect_chain = cls.get_redirect_chain(work_key)
+        summary['redirect_chain'] = [
+            {
+                "key": thing.key,
+                "occurrences": {},
+                "updates": {}
+            } for thing in redirect_chain
+        ]
+        summary['resolved_key'] = redirect_chain[-1].key
+
+        for r in summary['redirect_chain']:
+            olid = r['key'].split('/')[-1][2:-1]  # 'OL1234x' --> '1234'
+            new_olid = summary['resolved_key'].split('/')[-1][2:-1]
+
+            # count reading log entries
+            r['occurrences']['readinglog'] = len(
+                Bookshelves.get_works_shelves(olid))
+            r['occurrences']['ratings'] = len(
+                Ratings.get_all_works_ratings(olid))
+            r['occurrences']['booknotes'] = len(
+                Booknotes.get_booknotes_for_work(olid))
+            r['occurrences']['observations'] = len(
+                Observations.get_observations_for_work(olid))
+
+            # track updates
+            r['updates']['readinglog'] = Bookshelves.update_work_id(
+                olid, new_olid, _test=test
+            )
+            r['updates']['ratings'] = Ratings.update_work_id(
+                olid, new_olid, _test=test
+            )
+            r['updates']['booknotes'] = Booknotes.update_work_id(
+                olid, new_olid, _test=test
+            )
+            r['updates']['observations'] = Observations.update_work_id(
+                olid, new_olid, _test=test
+            )
+        return summary
+
+    @classmethod
+    def resolve_redirects_bulk(
+            cls,
+            batch_size=1000,
+            start_offset=0,
+            max_limit=1000,
+            grace_period_days=7,
+            cutoff_date=datetime.datetime(year=2017, month=1, day=1),
+            test=True
+    ):
+        """
+        batch_size - how many records to fetch per batch
+        start_offset - what offset to start from
+        pages - how many pages of batch_size to process
+        grace_period_days - ignore redirects created within period of days
+        cutoff_date - ignore redirects created before this date
+        test - don't resolve stale redirects, just identify them
+        """
+        pos = 0
+        grace_date = (
```

Comment:
This might be clearer as `max_date`, in that we've basically just got a date range to process.